clang/lib/CodeGen/CGExprAgg.cpp

```diff
@@ -536,6 +536,8 @@ void AggExprEmitter::EmitArrayInit(Address DestPtr, llvm::ArrayType *AType,
     CodeGen::CodeGenModule &CGM = CGF.CGM;
     ConstantEmitter Emitter(CGF);
     LangAS AS = ArrayQTy.getAddressSpace();
+    if (CGF.getLangOpts().OpenCL)
+      AS = LangAS::opencl_constant;
     if (llvm::Constant *C =
             Emitter.tryEmitForInitializer(ExprToVisit, AS, ArrayQTy)) {
       auto GV = new llvm::GlobalVariable(
```

Comment:
In the call to MakeAddrLValue, do we need to fix the address-space of ArrayQTy?  Maybe it happens to work at the moment because we infer the address-space for the memcpy itself from the LLVM pointer type.  But I'd prefer to fix it properly in case it ends up mattering in the future.

Reply:
It seems to have worked indeed, but we should probably adjust the address space of `ArrayQTy` everywhere then?  I'll upload a new revision to do so.

Reply:
The first argument EmitFinalDestCopy is the type of the destination, not the source.

Reply:
Nice catch, that wasn't obvious to me from the code.  Updated by splitting out a separate QualType variable for the global.